### PR TITLE
Adding a progress-bar widget

### PR DESCRIPTION
**Pull request by Weibye**
# Objective

- On our way to solve #5604 we need a lot of different widgets. 
- This attempts to add a progress-bar widget
- Which can be used for loading indicators, health / mana-bars.

## Solution

https://user-images.githubusercontent.com/13300393/200424470-89ad311c-fd54-482c-a984-5dab902c8929.mp4

- Same as #6236, the general design-approach here is "Widgets as lose collections of hierarchies of bundles".
- For the loading-bar widget to work, the user needs to create a `NodeBundle` with a `ProgressBarWidget`-component attached, that has a child `NodeBundle` with a `ProgressBarInner`-component. 
- It is completely up to the user to make sure these requirements are met.

---

### Added
- Progress bar widget

---
## Notes
- Please critique this general approach. I know how we can setup a lot of other widgets the same way, but I have not yet any clear answers for the bigger-picture problems such as "widget composition" and "widget styling".
- My gut feeling says it will be easier to tackle those problems once we have some real uses of various widgets in Bevy.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
@bzm3r review plz!

File: crates/bevy_ui/src/widget/loading_bar.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+};
+use bevy_hierarchy::{Children, Parent};
+use bevy_log::warn;
+use bevy_render::view::{ComputedVisibility, Visibility};
+use bevy_transform::prelude::{GlobalTransform, Transform};
+
+use crate::{BackgroundColor, FocusPolicy, Node, Size, Style, Val, ZIndex};
+
+#[derive(Component, Default, Clone, Debug)]
+pub struct LoadingBarWidget {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/widget/loading_bar.rs:17]: Doc string will help.

File: crates/bevy_ui/src/widget/loading_bar.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    progress: f32,
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/widget/loading_bar.rs:18]: I think you want a current, min and max value.

File: crates/bevy_ui/src/widget/loading_bar.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
+//! A loading bar widget.
+//! Can be used for loading bars, but also health-bars, mana, those kind of things.
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/widget/loading_bar.rs:2]: I think we should rename this to make that more obvious.

File: crates/bevy_ui/src/widget/loading_bar.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+pub(crate) fn update_loading_bars(
+    q: Query<(&LoadingBarWidget, &Children)>,
+    mut inner: Query<&mut Style, With<LoadingBarInner>>,
+) {
+    for (widget, children) in q.iter() {
+        for child in children.iter() {
+            if let Ok(mut style) = inner.get_mut(*child) {
+                style.size = Size::new(
+                    Val::Percent(widget.get_progress() * 100.0),
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/widget/loading_bar.rs:86]: What if I want a vertical bar?

File: examples/ui/loading_bar_widget.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        .spawn(NodeBundle {
+            style: Style {
+                size: Size::new(Val::Percent(100.0), Val::Percent(100.0)),
+                // horizontally center children nodes
+                justify_content: JustifyContent::Center,
+                // vertically center children nodes
+                align_items: AlignItems::Center,
+                ..default()
+            },
+            background_color: Color::rgba(19.0 / 255.0, 21.0 / 255.0, 22.0 / 255.0, 1.0).into(),
```
Comment (alice-i-cecile) [inline examples/ui/loading_bar_widget.rs:42]: These colors will be much clearer as a constant.

File: examples/ui/loading_bar_widget.rs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+                    ..default()
+                });
+            });
+        });
+}
+
+const LOAD_DURATION: f32 = 3.0;
+const COMPLETE_DURATION: f32 = 1.5;
+
+/// This is just a helper-system.
```
Comment (alice-i-cecile) [inline examples/ui/loading_bar_widget.rs:96]: This comment needs to be improved.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Specific feedback about this widget:

- genuinely useful
- needs to be more robust to misconfiguration and alternate use cases: vertical orientation, LTR, etc
- needs a name that encourages devs to use it as e.g. a life bar. `ProgressBar` maybe?
- needs an example demonstrating how to use and position it in screen space (yeah I know this currently sucks)

General thoughts:
- the "spawn complex trees of widgets" pattern is miserable. So much boilerplate, so hard to parse
- updating the value of data within components is great
- I worry about the poor scaling of O(n) change detection systems every frame for elements that rarely change. A push-based model feels more natural for thing like buttons
- differentiating widgets of the same type is going to get annoying without general-purpose relations unless you use an event-based architecture (which is annoying in its own ways)
- widgets really feel like they belong in their own `bevy_widgets` crate?
- I really want to be able to disable / not include systems for widgets I'm not using in my game: the random perf loss every single frame bothers me

File: examples/ui/loading_bar_widget.rs (hunk 5 above)
Comment (rparrett) [reply examples/ui/loading_bar_widget.rs:42]: [`rgba_u8`](https://docs.rs/bevy/latest/bevy/prelude/enum.Color.html#method.rgba_u8) would help

Comment (alice-i-cecile) [reply examples/ui/loading_bar_widget.rs:42]: Yep, but also the named constant will help explain either their color or their purpose.

**Comment by Weibye** [comment]
> widgets really feel like they belong in their own bevy_widgets crate?

We should also setup a new category of examples for widgets IMO. Should these two things happen in this PR, or somewhere else?

**Comment by alice-i-cecile** [comment]
I think in this PR: it's already going to be controversial, so let's do this right from the start.

File: examples/widgets/progress_bar.rs [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    let canvas_background: Color = Color::rgba_u8(19, 21, 22, 255);
+    let progress_bar_background: Color = Color::rgba_u8(29, 31, 33, 255);
+    let progress_bar_foreground: Color = Color::rgba_u8(50, 104, 159, 255);
+    let text_color: Color = Color::rgba_u8(197, 198, 190, 255);
+    let font = asset_server.load("fonts/FiraMono-Medium.ttf");
+
+    // ui camera
+    commands.spawn(Camera2dBundle::default());
+
+    // background that fills the entire viewport
```
Comment (IceSentry) [inline examples/widgets/progress_bar.rs:33]: Is this necessary? Why not just use the default clear color of bevy? I'm not really against this, but just trying to see where the example could be simplified a little.

Comment (IceSentry) [reply examples/widgets/progress_bar.rs:33]: I guess you need the root node to take the whole screen to center the progress bar?

Comment (Weibye) [reply examples/widgets/progress_bar.rs:33]: Yes. Previously we have been using auto-margin to centre elements on the page. This is not the best practice, see #6535. In order to centre the element you need the parent element to cover the whole area. 

The root / background does not strictly need a background colour, I added it for stylistic choices but can remove it if desired.

**Comment by AxiomaticSemantics** [comment]
I'd like to see support for generics for min, max, progress values if that is reasonable. (Most of my use of progress bars are based on {u,i}N values)
Another feature in my current hack to provide such a widget is bar draw direction partially noted above by Alice WRT horizontal and vertical drawing, but I also account for right and left sided origins.

**Comment by Weibye** [comment]
> I'd like to see support for generics for min, max, progress values if that is reasonable. (Most of my use of progress bars are based on {u,i}N values)

I'm not too versed in generic rust yet, so I'm not entirely sure how that would look but if you'd provide a code snippet of what that would look like, I'm happy to include it :)

File: crates/bevy_widget/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+pub use progress_bar::*;
+
+/// The basic plugin for Bevy Widget
+#[derive(Default)]
+pub struct WidgetPlugin;
+
+impl Plugin for WidgetPlugin {
+    fn build(&self, app: &mut App) {
+        app.register_type::<ProgressBarWidget>()
```
Comment (bzm3r) [inline crates/bevy_widget/src/lib.rs:16]: I'm not sure about `ProgressBar` as a name. The core widget here seems to be much more general than just a bar for progress.

A rough idea for a more general/useful name:  `ResourceMeter`? Not sold on that idea, but it is more general?

**Review by bzm3r — APPROVED** [review]
I like this.

File: crates/bevy_widget/src/lib.rs (hunk 8 above)
Comment (Weibye) [reply crates/bevy_widget/src/lib.rs:16]: In general UI terminology, the name of the widget is usually "ProgressBar". 

- https://material.angular.io/components/progress-bar/overview
- https://react-bootstrap.github.io/components/progress/
- https://docs.unity3d.com/ScriptReference/UIElements.ProgressBar.html
- https://docs.unrealengine.com/4.27/en-US/API/Runtime/UMG/Components/UProgressBar/
- https://docs.godotengine.org/en/stable/classes/class_progressbar.html

A loading screen, a health bar, mana bar, XP bar, various resource trackers you find in games are just specific implementations of that widget.

File: crates/bevy_widget/src/progress_bar.rs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+/// Marker component for the inner box of the progress bar.
+#[derive(Component, Default, Clone, Debug, Reflect)]
+#[reflect(Component)]
+pub struct ProgressBarInner;
+
+impl ProgressBarWidget {
+    /// Creates a new [`ProgressBarWidget`].
+    pub fn new(progress: f32, min: f32, max: f32) -> Self {
+        if min > max {
+            panic!("Min should not be larger than max");
```
[the remark this answers is not in the document]
Comment (Weibye) [reply crates/bevy_widget/src/progress_bar.rs:53]: Yeah that might be a better option

File: crates/bevy_math/src/lib.rs [hunk 10]
```diff
@@ -22,3 +22,8 @@ pub mod prelude {
 }
 
 pub use glam::*;
+
+/// Maps a value from one range of values to a new range of values.
+pub fn map_range(value: f32, old_range: (f32, f32), new_range: (f32, f32)) -> f32 {
+    (value - old_range.0) / (old_range.1 - old_range.0) * (new_range.1 - new_range.0) + new_range.0
+}
```
[the remark this answers is not in the document]
Comment (SolarLiner) [reply crates/bevy_math/src/lib.rs:29]: > How do we make this discoverable for others down the line?

Perhaps as an extension to either scalars or `Range` values, such that we can have either `value.map_range(a..b, c..d)` or `(a..b).map_range(c..d, value)`? In the former case the intervals are more intuitive (`a..b` is clearly the "old" range of `value`), whereas in the latter the arguments would be reversed, such that  `a..b` becomes the output range.

**Comment by richchurcher** [comment]
Backlog cleanup: closing due to inactivity, and in light of the significant development effort surrounding Bevy editor which will be surfacing any amount of new UI/widgets.
